File: typescript/examples/telemetry/03_sampling_and_backpressure.ts
// SPDX-FileCopyrightText: Copyright (C) 2026 MindTenet LLC
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of Undef Telemetry.

/**
 * 🎲 Sampling policies and backpressure queue controls.
 *
 * Demonstrates:
 * - SamplingPolicy with defaultRate and per-key overrides
 * - setSamplingPolicy / getSamplingPolicy / shouldSample
 * - QueuePolicy with per-signal maxSize limits
 * - setQueuePolicy / getQueuePolicy / tryAcquire / release
 * - getHealthSnapshot: dropped counts
 *
 * Run: npx tsx examples/telemetry/03_sampling_and_backpressure.ts
 */

import {
  counter,
  getHealthSnapshot,
  getLogger,
  getQueuePolicy,
  getSamplingPolicy,
  release,
  setQueuePolicy,
  setSamplingPolicy,
  setupTelemetry,
  shouldSample,
  shutdownTelemetry,
  tryAcquire,
  withTrace,
} from '../../src/index.js';

async function tracedWork(taskId: number): Promise<void> {
  await withTrace('example.sampling.concurrent', async () => {
    await new Promise((r) => setTimeout(r, 15));
    counter('example.sampling.counter').add(1, { task_id: String(taskId) });
  });
}

async function run(): Promise<void> {
  const log = getLogger('examples.sampling');

  // ── 🎲 Sampling policies with overrides ─────────────────
  console.log('🎲 Setting sampling policies...');
  setSamplingPolicy({ defaultRate: 0.0, overrides: { 'example.critical': 1.0 } });

  const logsPolicy = getSamplingPolicy();
  console.log(`  📋 defaultRate=${logsPolicy.defaultRate}, overrides=${JSON.stringify(logsPolicy.overrides)}`);

  // ── 🎯 shouldSample with overrides ─────────────────────
  console.log('\n🎯 shouldSample() decisions:');
  for (const key of ['example.routine', 'example.critical']) {
    const sampled = shouldSample(key);
    const icon = sampled ? '✅' : '❌';
    console.log(`  ${icon} ${key}: sampled=${sampled}`);
  }

  // Reset to full sampling for the work below
  setSamplingPolicy({ defaultRate: 1.0 });

  // ── 🚧 Backpressure queue limits ────────────────────────
  console.log('\n🚧 Setting queue policy (maxTraces=1)...');
  setQueuePolicy({ maxLogs: 0, maxMetrics: 0, maxTraces: 1 });
  const qp = getQueuePolicy();
  console.log(`  📋 Queue policy: logs=${qp.maxLogs}, traces=${qp.maxTraces}, metrics=${qp.maxMetrics}`);

  // ── tryAcquire / release demo ───────────────────────────
  console.log('\n🔑 tryAcquire / release demo (maxTraces=1):');
  const ticket1 = tryAcquire('traces');
  const ticket2 = tryAcquire('traces'); // should be null — at capacity
  console.log(`  ticket1=${ticket1 !== null ? `token=${ticket1.token}` : 'null'}`);
  console.log(`  ticket2=${ticket2 !== null ? `token=${ticket2.token}` : 'null (backpressure applied)'}`);
  if (ticket1) release(ticket1);
  console.log('  released ticket1');
  const ticket3 = tryAcquire('traces'); // should succeed now
  console.log(`  ticket3=${ticket3 !== null ? `token=${ticket3.token} (slot freed)` : 'null'}`);
  if (ticket3) release(ticket3);

  // ── ⚡ Concurrent traced work ──────────────────────────
  console.log('\n⚡ Launching 5 concurrent traced tasks...');
  setQueuePolicy({ maxLogs: 0, maxMetrics: 0, maxTraces: 0 }); // unlimited
  await Promise.all(Array.from({ length: 5 }, (_, i) => tracedWork(i)));
  console.log('  ✅ All tasks completed');

  // This event is sampled out (rate=0 override was reset, logs rate=1 now).
  log.info({ event: 'example.sampling.done' });

  // ── 📊 Health snapshot ──────────────────────────────────
  console.log('\n📊 Health snapshot:');
  const snapshot = getHealthSnapshot();
  console.log(`  📉 logsDropped:    ${snapshot.logsDropped}`);
  console.log(`  📉 tracesDropped:  ${snapshot.tracesDropped}`);
  console.log(`  📉 metricsDropped: ${snapshot.metricsDropped}`);

  console.log('\n🏁 Done!');
}

async function main(): Promise<void> {
  console.log('🎲 Sampling & Backpressure Demo\n');
  setupTelemetry({ serviceName: 'undef-telemetry-examples', consoleOutput: false });
  await run();
  await shutdownTelemetry();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/telemetry/03_sampling_and_backpressure.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

"""🎲 Sampling policies and backpressure queue controls.

Demonstrates:
- SamplingPolicy with default_rate and per-key overrides
- set_sampling_policy / get_sampling_policy / should_sample
- QueuePolicy with per-signal maxsize
- set_queue_policy / get_queue_policy
- HealthSnapshot: dropped counts and queue depths
"""

from __future__ import annotations

import asyncio

from provide.telemetry import (
    QueuePolicy,
    SamplingPolicy,
    counter,
    event,
    get_health_snapshot,
    get_logger,
    get_queue_policy,
    get_sampling_policy,
    set_queue_policy,
    set_sampling_policy,
    setup_telemetry,
    should_sample,
    shutdown_telemetry,
    trace,
)


@trace(event("example", "sampling", "concurrent"))
async def _traced_work(task_id: int) -> None:
    await asyncio.sleep(0.15)
    counter("example.sampling.counter").add(1, {"task_id": str(task_id)})


async def _run() -> None:
    log = get_logger("examples.sampling")

    # ── 🎲 Sampling policies with overrides ─────────────────
    print("🎲 Setting sampling policies...")
    set_sampling_policy(
        "logs",
        SamplingPolicy(default_rate=0.0, overrides={"example.critical": 1.0}),
    )
    set_sampling_policy("metrics", SamplingPolicy(default_rate=1.0))
    set_sampling_policy("traces", SamplingPolicy(default_rate=1.0))

    # ── 🔍 Inspect active policies ──────────────────────────
    logs_policy = get_sampling_policy("logs")
    print(f"  📋 logs:    default_rate={logs_policy.default_rate}, overrides={logs_policy.overrides}")
    print(f"  📋 metrics: default_rate={get_sampling_policy('metrics').default_rate}")
    print(f"  📋 traces:  default_rate={get_sampling_policy('traces').default_rate}")

    # ── 🎯 should_sample with overrides ─────────────────────
    print("\n🎯 should_sample() decisions:")
    for key in ("example.routine", "example.critical"):
        sampled = should_sample("logs", key=key)
        icon = "✅" if sampled else "❌"
        print(f"  {icon} logs/{key}: sampled={sampled}")

    # ── 🚧 Backpressure queue limits ────────────────────────
    print("\n🚧 Setting queue policy (traces_maxsize=1)...")
    set_queue_policy(QueuePolicy(logs_maxsize=0, metrics_maxsize=0, traces_maxsize=1))
    qp = get_queue_policy()
    print(f"  📋 Queue policy: logs={qp.logs_maxsize}, traces={qp.traces_maxsize}, metrics={qp.metrics_maxsize}")

    # ── ⚡ Concurrent traced work (will saturate queue) ─────
    print("\n⚡ Launching 5 concurrent traced tasks...")
    tasks = [asyncio.create_task(_traced_work(i)) for i in range(5)]
    await asyncio.gather(*tasks)
    print("  ✅ All tasks completed")

    # This event itself is sampled out (logs rate=0%).
    log.info(event("example", "sampling", "done"))

    # ── 📊 Health snapshot ──────────────────────────────────
    print("\n📊 Health snapshot after saturation:")
    snapshot = get_health_snapshot()
    print(f"  📉 dropped_logs:         {snapshot.dropped_logs}")
    print(f"  📉 dropped_traces:       {snapshot.dropped_traces}")
    print(f"  📉 dropped_metrics:      {snapshot.dropped_metrics}")
    print(f"  📤 export_failures_traces: {snapshot.export_failures_traces}")

    print("\n🏁 Done!")


def main() -> None:
    print("🎲 Sampling & Backpressure Demo\n")
    setup_telemetry()
    asyncio.run(_run())
    shutdown_telemetry()


if __name__ == "__main__":
    main()
